### PR TITLE
feat(ingest/BigQuery): refactor+parallelize dataset metadata extraction

### PR DESCRIPTION
**Pull request by mayurinehate**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

## Summary by CodeRabbit

- **New Features**
  - Introduced a parallel processing configuration for BigQuery schemas to enhance performance.
  - Added `BigQueryTestConnection` class for improved connection testing and capability reporting.

- **Improvements**
  - Updated BigQuery and Snowflake ingestion modules for better error handling, logging, and API request tracking.
  - Enhanced BigQuery schema generator and metadata extraction processes for efficiency and reliability.
  - Increased default batch processing limits for BigQuery datasets.

- **Bug Fixes**
  - Corrected warning and error message structures to provide clearer information.

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema_gen.py [hunk 1]
```diff
@@ -0,0 +1,1096 @@
...
+                trace = traceback.format_exc()
+                logger.error(trace)
+                logger.error(error_message, exc_info=e)
+                self.report.report_failure(
+                    "metadata-extraction",
+                    f"{project_id}.{bigquery_dataset.name} - {error_message} - {trace}",
+                )
+
+        with concurrent.futures.ThreadPoolExecutor(
+            max_workers=SCHEMA_PARALLELISM
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema_gen.py:370]: let's pass this in via a config variable

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema_gen.py [hunk 2]
```diff
@@ -0,0 +1,1096 @@
...
+                        pass
+
+                # Filter out the done futures.
+                futures = [f for f in futures if not f.done()]
+                if not futures:
+                    break
+
+        # Yield the remaining work units. This theoretically should not happen, but adding it just in case.
+        while not q.empty():
+            yield q.get_nowait()
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema_gen.py:395]: does it make sense to pull some of this out into a helper method or something?

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema_gen.py [hunk 3]
```diff
@@ -0,0 +1,1096 @@
...
+            except Exception as e:
+                error_message = f"Unable to get tables for dataset {bigquery_dataset.name} in project {project_id}, skipping. Does your service account has bigquery.tables.list, bigquery.routines.get, bigquery.routines.list permission? The error was: {e}"
+                if self.config.is_profiling_enabled():
+                    error_message = f"Unable to get tables for dataset {bigquery_dataset.name} in project {project_id}, skipping. Does your service account has bigquery.tables.list, bigquery.routines.get, bigquery.routines.list permission, bigquery.tables.getData permission? The error was: {e}"
+                trace = traceback.format_exc()
+                logger.error(trace)
+                logger.error(error_message, exc_info=e)
+                self.report.report_failure(
+                    "metadata-extraction",
+                    f"{project_id}.{bigquery_dataset.name} - {error_message} - {trace}",
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema_gen.py:366]: tweak error reporting here

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema_gen.py [hunk 4]
```diff
@@ -0,0 +1,1096 @@
...
+                ):
+                    q.put(wu)
+            except Exception as e:
+                error_message = f"Unable to get tables for dataset {bigquery_dataset.name} in project {project_id}, skipping. Does your service account has bigquery.tables.list, bigquery.routines.get, bigquery.routines.list permission? The error was: {e}"
+                if self.config.is_profiling_enabled():
+                    error_message = f"Unable to get tables for dataset {bigquery_dataset.name} in project {project_id}, skipping. Does your service account has bigquery.tables.list, bigquery.routines.get, bigquery.routines.list permission, bigquery.tables.getData permission? The error was: {e}"
+                trace = traceback.format_exc()
+                logger.error(trace)
+                logger.error(error_message, exc_info=e)
+                self.report.report_failure(
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema_gen.py:364]:
```suggestion
                self.report.failure(
```

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema_gen.py [hunk 5]
```diff
@@ -0,0 +1,1096 @@
...
+from datahub.utilities.registries.domain_registry import DomainRegistry
+
+logger: logging.Logger = logging.getLogger(__name__)
+# Handle table snapshots
+# See https://cloud.google.com/bigquery/docs/table-snapshots-intro.
+SNAPSHOT_TABLE_REGEX = re.compile(r"^(.+)@(\d{13})$")
+CLUSTERING_COLUMN_TAG = "CLUSTERING_COLUMN"
+
+
+class BigQueryProjectSchemaGenerator:
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema_gen.py:111]: is this specific to a single project?

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema_gen.py (hunk 2 above)
Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema_gen.py:395]: ideally reused across bq and snowflake

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema_gen.py (hunk 5 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema_gen.py:110]: Renaming this to `BigQuerySchemaGenerator`. Currently calling its `get_project_workunits` for every project. It maintains the cache of entity references across all projects that it has processed.
```
       self.bq_schema_extractor.view_refs_by_project,
        self.bq_schema_extractor.view_definitions,
        self.bq_schema_extractor.snapshot_refs_by_project,
        self.bq_schema_extractor.snapshots_by_ref,
        self.bq_schema_extractor.table_refs,
```

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema_gen.py (hunk 2 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema_gen.py:409]: Yes

Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema_gen.py]: Done

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_config.py [hunk 6]
```diff
@@ -24,6 +24,8 @@
 
 logger = logging.getLogger(__name__)
 
+SCHEMA_PARALLELISM = int(os.getenv("DATAHUB_BIGQUERY_SCHEMA_PARALLELISM", 20))
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_config.py:27]: just to make things more clear

```suggestion
DEFAULT_BQ_SCHEMA_PARALLELISM = int(os.getenv("DATAHUB_BIGQUERY_SCHEMA_PARALLELISM", 20))
```

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py [hunk 7]
```diff
@@ -163,33 +164,31 @@ def _should_retry(exc: BaseException) -> bool:
...
+                retry=retry.Retry(
+                    predicate=_should_retry, initial=10, maximum=180, timeout=900
                 )
-                projects: List[BigqueryProject] = [
-                    BigqueryProject(id=p.project_id, name=p.friendly_name)
-                    for p in projects_iterator
-                ]
-                self.report.num_list_projects = len(projects)
-                return projects
-            except Exception as e:
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py:185]: is this exception handled elsewhere?

**Review by hsheth2 — APPROVED** [review]

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py (hunk 7 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_schema.py:185]: Yes, this is moved up to caller of this method , where caller catches and reports error along with failure.

**Pull request merged**